### DISCUSSION TITLE
Why put static and dynamic libraries in the same directory?

### DISCUSSION THREAD

**Discussion by wAuner** [Homebrew/discussions#4672] [Everyday usage]
### Output of `brew config`

```shell
not relevant
```

### Output of `brew doctor`

```shell
not relevant
```

### Description of issue

If a library like `fmt` provides dynamic `.dylib` and static `.a` binaries, brew puts them side-by-side in the same directory.
```
.
├── ChangeLog.rst
├── include
│  └── fmt
├── INSTALL_RECEIPT.json
├── lib
│  ├── cmake
│  ├── libfmt.10.0.0.dylib
│  ├── libfmt.10.dylib -> libfmt.10.0.0.dylib
│  ├── libfmt.a
│  ├── libfmt.dylib -> libfmt.10.dylib
│  └── pkgconfig
├── LICENSE.rst
└── README.rst
```

However, this is a HUGE pain when using clang and especially Xcode. clang's cli (or rather ld) will always link the dynamic library when specified via `-lfmt` as long as both live in the same directory. 
Xcode is even worse. Even if you add specifically the `.a` file to the linking step, it still chooses the dynamic lib in this case. The only clean way to get the lib linked statically I see, is to separate libraries into static and dynamic subdirs. 
Is there a reason that this is not done? How are others dealing with this?

Things get even worse if you adopt the hardened runtime on macOS (which Xcode does by default). Then you'll have to manually sign all the brew libraries if they're linked dynamically or add an exception entitlement, which is not allowed for system extensions. That's one reason why I much prefer to link statically most of the time.

**Comment by SMillerDev** [comment]
As far as I know this is the standard for unix. That's why Homebrew uses it.

**Reply by wAuner** [reply]
right, but on macOS this really doesn't work with the available tooling. On Linux for example with gcc and the gnu linker you're able to properly specify what you want by using the syntax `-l:libfmt.a` instead of `-lfmt`. This is not supported by (Apple) clang.

**Reply by SMillerDev** [reply]
That seems like something that would be good to file with Apple as a bug report. Either way, static linking isn't something Homebrew is optimising for.

**Reply by wAuner** [reply]
I've already filed a report with Apple for the issue. I don't think Homebrew should optimize for static linking, but at least enable the possibility. Otherwise why include them in the first place, if it's not (reasonably) possible to use them?
A solution could be to place static builds in a `static/` subdir. This wouldn't change the current behavior, because dynamic libs are linked as before. but allow static linking.

**Reply by gromgit** [reply]
Not a macOS developer, but in my *nix developer days, I'd say `/path/to/libfmt.a` if I wanted static linking, and `-L/path/to -lfmt` if I wanted dynamic linking, _but never both at the same time_. You might want to check your project's generated build commands to see if that's what is happening here.

**Reply by wAuner** [reply]
@gromgit It's true that it's possible to link statically by including the full path to the archive without a linker flag, like `clang++ libfmt.a main.cpp`. This however is only possible via a hack with Xcode and makes dependency management a nightmare.
But I don't know what mean by "but never both at the same time". 
The "normal" way (as per clang documentation and what other package managers like conan are doing) for both static and dynamic linking is to use `-L` and `-l`. With clang you have to to put dylib and static libs in different directories so that you can pass them via `-L`.

**Reply by gromgit** [reply]
> But I don't know what mean by "but never both at the same time".

As in "I'd never use both forms for the same library in the same command line", but with build environments more complex that `make`, I can see people fighting their own tools to get the exact build commands they want. :man_shrugging:

**Reply by wAuner** [reply]
"I'd never use both forms for the same library in the same command line" ? I don't see why anyone would do that or how this is related.

**Reply by gromgit** [reply]
No one would do this deliberately, but when you use complex build tools with configuration languages and methodologies far removed from the final build commands, it's possible for the tools to do this kind of boneheaded thing for you _unintentionally_, especially since tool developers generally optimize for simple use cases.

Anyway, this is getting rather far from the original point, which is that _by convention_, most software projects dump their static and dynamic libraries in the same directory, and many project build configuration mechanisms (e.g. the ever-popular GNU Autoconf) don't explicitly support separate directories thereof. Even Apple clang's toolchain seems to do the same thing, while naming their dylibs `libXXX_dynamic.dylib` to presumably help `ld` distinguish them from the `libXXX.a`s in the same directory.

For Homebrew to move static libs to separate directories, or add distinguishing tags to said libraries, would require changes across all formulae that export static libraries, which is quite an undertaking.
